File: backend/app/common/deliverables.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class DeliverableDefinition:
    code: str
    platform: str
    legacy_type: str
    label: str


DELIVERABLE_DEFINITIONS: dict[str, DeliverableDefinition] = {
    "youtube_live": DeliverableDefinition("youtube_live", "youtube", "live_stream", "YouTube Live"),
    "youtube_short": DeliverableDefinition("youtube_short", "youtube", "short_video", "YouTube Shorts"),
    "youtube_video": DeliverableDefinition("youtube_video", "youtube", "dedicated_video", "YouTube Video"),
    "instagram_live": DeliverableDefinition("instagram_live", "instagram", "live_stream", "Instagram Live"),
    "instagram_feed": DeliverableDefinition("instagram_feed", "instagram", "photo_post", "Instagram Feed"),
    "instagram_reel": DeliverableDefinition("instagram_reel", "instagram", "short_video", "Instagram Reel"),
    "instagram_story": DeliverableDefinition("instagram_story", "instagram", "story", "Instagram Story"),
    "tiktok_live": DeliverableDefinition("tiktok_live", "tiktok", "live_stream", "TikTok Live"),
    "tiktok_story": DeliverableDefinition("tiktok_story", "tiktok", "story", "TikTok Story"),
    "tiktok_video": DeliverableDefinition("tiktok_video", "tiktok", "short_video", "TikTok Video"),
}

LEGACY_TO_CANONICAL: dict[tuple[str, str], str] = {
    ("youtube", "live_stream"): "youtube_live",
    ("youtube", "short_video"): "youtube_short",
    ("youtube", "dedicated_video"): "youtube_video",
    ("youtube", "integrated_mention"): "youtube_video",
    ("instagram", "live_stream"): "instagram_live",
    ("instagram", "photo_post"): "instagram_feed",
    ("instagram", "short_video"): "instagram_reel",
    ("instagram", "story"): "instagram_story",
    ("tiktok", "live_stream"): "tiktok_live",
    ("tiktok", "story"): "tiktok_story",
    ("tiktok", "short_video"): "tiktok_video",
}


def canonical_deliverable_code(
    *,
    platform: str | None,
    deliverable_code: str | None = None,
    legacy_type: str | None = None,
) -> str | None:
    if deliverable_code:
        return deliverable_code
    if platform and legacy_type:
        return LEGACY_TO_CANONICAL.get((platform, legacy_type))
    return None


def legacy_deliverable_type(
    *,
    deliverable_code: str | None = None,
    platform: str | None = None,
    legacy_type: str | None = None,
) -> str | None:
    if deliverable_code:
        definition = DELIVERABLE_DEFINITIONS.get(deliverable_code)
        if definition:
            return definition.legacy_type
    return legacy_type


def deliverable_platform(
    *,
    deliverable_code: str | None = None,
    platform: str | None = None,
) -> str | None:
    if deliverable_code:
        definition = DELIVERABLE_DEFINITIONS.get(deliverable_code)
        if definition:
            return definition.platform
    return platform

File: frontend/cohesiq-v0/lib/deliverables.ts
import type { DeliverableCode, DeliverableType, PlatformType } from "./types";

export const DELIVERABLE_DEFINITIONS: Record<DeliverableCode, {
  code: DeliverableCode;
  platform: PlatformType;
  legacyType: DeliverableType;
  label: string;
}> = {
  youtube_live: { code: "youtube_live", platform: "youtube", legacyType: "live_stream", label: "YouTube Live" },
  youtube_short: { code: "youtube_short", platform: "youtube", legacyType: "short_video", label: "YouTube Shorts" },
  youtube_video: { code: "youtube_video", platform: "youtube", legacyType: "dedicated_video", label: "YouTube Video" },
  instagram_live: { code: "instagram_live", platform: "instagram", legacyType: "live_stream", label: "Instagram Live" },
  instagram_feed: { code: "instagram_feed", platform: "instagram", legacyType: "photo_post", label: "Instagram Feed" },
  instagram_reel: { code: "instagram_reel", platform: "instagram", legacyType: "short_video", label: "Instagram Reel" },
  instagram_story: { code: "instagram_story", platform: "instagram", legacyType: "story", label: "Instagram Story" },
  tiktok_live: { code: "tiktok_live", platform: "tiktok", legacyType: "live_stream", label: "TikTok Live" },
  tiktok_story: { code: "tiktok_story", platform: "tiktok", legacyType: "story", label: "TikTok Story" },
  tiktok_video: { code: "tiktok_video", platform: "tiktok", legacyType: "short_video", label: "TikTok Video" },
};

const LEGACY_TO_CANONICAL: Partial<Record<`${PlatformType}:${DeliverableType}`, DeliverableCode>> = {
  "youtube:live_stream": "youtube_live",
  "youtube:short_video": "youtube_short",
  "youtube:dedicated_video": "youtube_video",
  "youtube:integrated_mention": "youtube_video",
  "instagram:live_stream": "instagram_live",
  "instagram:photo_post": "instagram_feed",
  "instagram:short_video": "instagram_reel",
  "instagram:story": "instagram_story",
  "tiktok:live_stream": "tiktok_live",
  "tiktok:story": "tiktok_story",
  "tiktok:short_video": "tiktok_video",
};

const LEGACY_LABELS: Record<DeliverableType, string> = {
  dedicated_video: "Dedicated Video",
  integrated_mention: "Integrated Mention",
  short_video: "Short Video",
  photo_post: "Photo Post",
  story: "Story",
  live_stream: "Live Stream",
  blog_post: "Blog Post",
  other: "Other",
};

export function resolveDeliverableCode(
  platform?: PlatformType,
  deliverableCode?: string | null,
  deliverableType?: DeliverableType | null,
): DeliverableCode | undefined {
  if (deliverableCode && deliverableCode in DELIVERABLE_DEFINITIONS) {
    return deliverableCode as DeliverableCode;
  }
  if (platform && deliverableType) {
    return LEGACY_TO_CANONICAL[`${platform}:${deliverableType}`];
  }
  return undefined;
}

export function getDeliverableLabel(
  platform?: PlatformType,
  deliverableCode?: string | null,
  deliverableType?: DeliverableType | null,
): string {
  const canonical = resolveDeliverableCode(platform, deliverableCode, deliverableType);
  if (canonical) {
    return DELIVERABLE_DEFINITIONS[canonical].label;
  }
  if (deliverableType) {
    return LEGACY_LABELS[deliverableType] ?? deliverableType;
  }
  return "Deliverable";
}
